src/include/storage/data_table.h

```diff
@@ -15,6 +15,12 @@ class TransactionManager;
 
 namespace terrier::storage {
 
+namespace index {
+class Index;
+template <typename KeyType>
+class BwTreeIndex;
```

Comment:
I can see about moving all of the necessary logic up to the Index level so the underlying wrappers don't need to talk to the DataTable.

Reply:
I'm not seeing a clean way to do this without modifying the underlying bwtree's predicate checking lambda to take variadic arguments (so I can pass the txn context). Right now it's captured implicitly by reference from the calling scope, but if I move the predicate function up to the wrapper level, then I need to add it as an argument and can't pass that function directly to the bwtree because it only expects 1 argument (the value type).

Reply:
Ok.  Maybe put a TODO or some other note here in case we end up replacing the BwTree with something else?

Reply:
I don't think there is there right place for it. I'll add more to the top of index.h describing how to add future indexes.